File: LP/LAB_PYTHON/LAB1/funcionsAmbLlistes.py
def myLength(L):
    r = 0
    for x in L:
        r = r + 1
    return r

def myMaximum(L):
    m = L[0]
    for elem in L:
        if elem > m:
            m = elem
    return m


def average(L):
    acu = 0
    for elem in L:
        acu = acu + elem
    return (acu/(len(L)))


def flipList(L):
    if myLength(L) == 0:
        return []
    else:
        actual = L.pop()
        return [actual] + flipList(L) 

    
def buildPalindrome(L):
    Linvertida = flipList(L[:])
    return  Linvertida + L


def flatten(L):
    if isinstance(L, list):
        if L != []:
            fParcial = flatten (L.pop())
            return (flatten(L) + fParcial)
        else:
            return []
    else:
        return [L]


def remove(L1, L2):
    r = []
    for elem in L1:
        if not elem in L2:
            r.append(elem)
    return r


def oddsNevens(L):
    impares = [i for i in L if i%2 != 0]
    pares = [i for i in L if i%2 == 0]
    t = (impares, pares)
    return t


def isPrime(x):
    divisors = [n for n in range (1, x+1) if x % n == 0]
    return divisors == [1, x]


def primeDivisors(n):
    if isPrime(n):
        return [n]
    return [i for i in range(1,n) if (n%i == 0) and isPrime(i)] 

File: LP/LAB_HASKELL/1_INIT/funcions_amb_llistes.hs
myLength :: [Int] -> Int
myLength [] = 0
myLength(_:xs) = 1 + myLength xs

myMaximum :: [Int] -> Int
myMaximum[a] = a
myMaximum(a:xs) 
    | myMaximum xs > a = myMaximum xs
    | otherwise = a

suma :: [Int] -> Int
suma[] = 0
suma(x:xs) = x + suma xs

average :: [Int] -> Float
average (xs) = fromIntegral (div (suma xs) (myLength xs))

buildPalindrome :: [Int] -> [Int]
buildPalindrome(xs) = reverse xs ++ xs

--si 
remove :: [Int] -> [Int] -> [Int]
remove a []= a
remove [] _ = []
remove (x:xs) b 
    | elem x b = remove xs b
    | otherwise = x : remove xs b

flatten :: [[Int]] -> [Int]
flatten [] = []
flatten (a:xs) = a ++ flatten xs


oddList :: [Int] -> [Int]
oddList [] = []
oddList (a:xs) 
    | mod a 2 == 0 = a : oddList xs
    | otherwise = oddList xs

evenList :: [Int] -> [Int]
evenList [] = []
evenList (a:xs)
    | mod a 2 /= 0 = a : evenList xs
    | otherwise = evenList xs

oddsNevens :: [Int] -> ([Int],[Int])
oddsNevens [] = ([],[])
oddsNevens xs = (evenList xs, oddList xs)

divisors :: Int -> [Int]
divisors n = [x | x <-[1..n], n `mod` x == 0]

isPrime :: Int -> Bool
isPrime n = divisors n == [1,n]

primeDivisors :: Int -> [Int]
primeDivisors n = [x | x <-[1..n], mod n x == 0, isPrime x]
